rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -1324,4 +1457,146 @@ size_t Player::get_number_of_registered_on_play_msg_post_callbacks()
 }
 
+PlayerImpl::PlayerClient::PlayerClient(
+  std::shared_ptr<rclcpp::GenericClient> cli,
+  std::string service_name,
+  std::string service_event_type,
+  const rclcpp::Logger & logger)
+: client_(std::move(cli)),
+  service_name_(service_name),
+  logger_(logger)
+{
+  ts_lib_ = rclcpp::get_typesupport_library(
+    service_event_type, "rosidl_typesupport_cpp");
+
+  ts_ = rclcpp::get_typesupport_handle(
+    service_event_type,
+    "rosidl_typesupport_cpp",
+    *ts_lib_);
+
+  auto ts_handle = get_message_typesupport_handle(
+    ts_,
+    rosidl_typesupport_introspection_cpp::typesupport_identifier);
+
+  message_members_ =
+    reinterpret_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+    ts_handle->data);
+}
+
+bool PlayerImpl::PlayerClient::is_include_request_message(const rclcpp::SerializedMessage & message)
+{
+  auto type = get_msg_event_type(message);
+
+  // Ignore response message
+  if (type == service_msgs::msg::ServiceEventInfo::RESPONSE_SENT ||
+    type == service_msgs::msg::ServiceEventInfo::RESPONSE_RECEIVED)
+  {
+    return false;
+  }
+
+  // Ignore metadata message
+  if (type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT &&
+    client_side_type_ == introspection_type::METADATA)
+  {
+    return false;
+  }
+  if (type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED &&
+    service_side_type_ == introspection_type::METADATA)
+  {
+    return false;
+  }
+
+  if (client_side_type_ == introspection_type::UNKNOW &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      client_side_type_ = introspection_type::METADATA;
+      RCLCPP_WARN(
+        logger_,
+        "The configuration of introspection for '%s' client is metadata !",
+        service_name_.c_str());
+      return false;
+    } else {
+      client_side_type_ = introspection_type::CONTENTS;
+    }
+  }
+  if (service_side_type_ == introspection_type::UNKNOW &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      service_side_type_ = introspection_type::METADATA;
+      RCLCPP_WARN(
+        logger_,
+        "The configuration of introspection for '%s' service is metadata !",
+        service_name_.c_str());
+      return false;
+    } else {
+      service_side_type_ = introspection_type::CONTENTS;
+    }
+  }
+
+  // If there are request send info and request receive info, only send request send info.
+  if (client_side_type_ == introspection_type::CONTENTS &&
+    service_side_type_ == introspection_type::CONTENTS &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+void PlayerImpl::PlayerClient::async_send_request(const rclcpp::SerializedMessage & message)
+{
```

Comment:
I have a concern about `PlayerClient::async_send_request`. We can’t really make it sending request asynchronously at least with the current design. Because we need to wait for future returning from `client_->async_send_request(..)`. Since we can't delete `ros_message` until the future complete. Otherwise, it will be a dangling pointer to the `request_addr` and memory could be reused and we have a chance to use corrupted memory during the `client_->async_send_request(..)`. Also according to the notes in the `client_→async_send_request(..)` API need to call `client_→remove_pending_request(future)` if we waited for future and timeout happened.